### PR TITLE
leverage new source-to-image API around git clone spec validation/correction

### PR DESCRIPTION
**Pull request by gabemontero**
@bparees @csrwng revision 2 of the origin side changes for leveraging the improved git clone spec validation.

PTAL - thx

### CODE REVIEW

File: pkg/generate/app/cmd/newapp_test.go [hunk 1]
```diff
@@ -66,8 +66,8 @@ func TestAddArguments(t *testing.T) {
 			unknown:    []string{},
 		},
 		"source": {
-			args:    []string{".", testDir, "git://server/repo.git"},
-			repos:   util.StringList([]string{".", testDir, "git://server/repo.git"}),
+			args:    []string{".", testDir, "git://github.com/openshift/origin.git"},
+			repos:   util.StringList([]string{".", testDir, "git://github.com/openshift/origin.git"}),
```
Comment (bparees) [inline pkg/generate/app/cmd/newapp_test.go]: if this test is going to end up cloning this repo, we should pick a smaller one.  origin is huge.

File: pkg/generate/app/cmd/newapp_test.go [hunk 2]
```diff
@@ -76,9 +76,9 @@ func TestAddArguments(t *testing.T) {
 			unknown: []string{},
 		},
 		"mix 1": {
-			args:       []string{"git://server/repo.git", "mysql+ruby~git@test.server/repo.git", "env1=test", "ruby-helloworld-sample"},
-			repos:      util.StringList([]string{"git://server/repo.git"}),
-			components: util.StringList([]string{"mysql+ruby~git@test.server/repo.git", "ruby-helloworld-sample"}),
+			args:       []string{"git://github.com/openshift/origin.git", "mysql+ruby~git@github.com/openshift/origin.git", "env1=test", "ruby-helloworld-sample"},
+			repos:      util.StringList([]string{"git://github.com/openshift/origin.git"}),
+			components: util.StringList([]string{"mysql+ruby~git@github.com/openshift/origin.git", "ruby-helloworld-sample"}),
```
Comment (bparees) [inline pkg/generate/app/cmd/newapp_test.go]: same for all of these.

File: pkg/generate/app/cmd/newapp_test.go [hunk 3]
```diff
@@ -1503,14 +1509,23 @@ func fakeSimpleDockerSearcher() app.Searcher {
...
 // testing
-func MockSourceRepositories(t *testing.T) []*app.SourceRepository {
+func MockSourceRepositories(t *testing.T, file string) []*app.SourceRepository {
 	var b []*app.SourceRepository
 	for _, location := range []string{
-		"some/location.git",
+		file,
 		"https://github.com/openshift/ruby-hello-world.git",
-		"another/location.git",
+		file,
```
Comment (bparees) [inline pkg/generate/app/cmd/newapp_test.go]: what's the value in having two identical repository paths in this array?

File: pkg/generate/app/cmd/newapp_test.go (hunk 2 above)
Comment (gabemontero) [inline pkg/generate/app/cmd/newapp_test.go]: actually, it is not a `git clone` but a `git ls-remote` ... it is lightweight/super fast

File: pkg/generate/git/repository.go [hunk 4]
```diff
@@ -147,6 +148,12 @@ func (r *repository) Clone(location string, url string) error {
 	return err
 }
 
+// LsRemote lists references in a remote repository
+func (r *repository) LsRemote(url string) error {
+	_, _, err := r.git(nil, "", "ls-remote", url)
```
Comment (mfojtik) [inline pkg/generate/git/repository.go]: https://github.com/openshift/origin/pull/5573

File: pkg/generate/app/cmd/newapp_test.go (hunk 3 above)
Comment (gabemontero) [inline pkg/generate/app/cmd/newapp_test.go]: correcting

File: pkg/generate/git/repository.go (hunk 4 above)
Comment (mfojtik) [inline pkg/generate/git/repository.go]: also maybe rename this func to `ListRemote` the `Ls` in name is weird

File: pkg/generate/app/sourcelookup.go [hunk 5]
```diff
@@ -13,11 +12,7 @@ import (
 	"github.com/openshift/origin/pkg/generate/dockerfile"
 	"github.com/openshift/origin/pkg/generate/git"
 	"github.com/openshift/origin/pkg/generate/source"
-)
-
-var (
-	argumentGit         = regexp.MustCompile("^(http://|https://|git@|git://).*(?:#([a-zA-Z0-9]*))?$")
-	argumentGitProtocol = regexp.MustCompile("^(git@|git://)")
+	s2igit "github.com/openshift/source-to-image/pkg/scm/git"
```
Comment (mfojtik) [inline pkg/generate/app/sourcelookup.go]: @bparees as we discussed earlier, we will get rid of git in s2i long term :-) All this code will has to move to origin then.

**Comment by gabemontero** [comment]
responses to comments pushed

**Comment by gabemontero** [comment]
[test]

**Comment by gabemontero** [comment]
fyi - the test failure above was e2e-docker ... failure to connect to a container ... seems unrelated to this change.

**Comment by gabemontero** [comment]
update for forcePull to pullPolicy pushed

File: pkg/build/builder/sti.go [hunk 6]
```diff
@@ -159,6 +159,12 @@ func (s *STIBuilder) Build() error {
 		ContextDir: s.build.Spec.Source.ContextDir,
 	}
 
+	if config.ForcePull {
+		config.PreviousImagePullPolicy = s2iapi.PullAlways
+	} else {
+		config.PreviousImagePullPolicy = s2iapi.PullIfNotPresent
+	}
+
```
Comment (bparees) [inline pkg/build/builder/sti.go]: shouldn't we also stop setting ForcePull above?

Comment (bparees) [inline pkg/build/builder/sti.go]: (and make this if be based on s.build.Spec.Strategy.SourceStrategy.ForcePull)

File: pkg/generate/app/cmd/newapp_test.go (hunk 2 above)
Comment (bparees) [inline pkg/generate/app/cmd/newapp_test.go]: ok.  wasn't sure where it was getting used.

**Comment by bparees** [comment]
one nit and lgtm.

**Comment by gabemontero** [comment]
update for force pull / pull policy nit pushed

**Comment by bparees** [comment]
lgtm, but this would probably be a good time to run your extended tests that exercise the forcepull flag :)

**Comment by gabemontero** [comment]
i ran all the extended tests locally, but let me try running the forcepull in jenkins now :-)

[testextendedonly][extended:forcepull]

**Comment by bparees** [comment]
[testonlyextended][extended:core(forcepull)]

**Comment by mfojtik** [comment]
hold on merge, will review this tomorrow

**Comment by gabemontero** [comment]
[testonlyextended][extended:core(forcepull)]

**Comment by gabemontero** [comment]
@bparees @mfojtik  - the forcepull extended test actually passed.  I'll paste a copule of snippets from the jenkins console:
1)  ok       builds: forcepull: ForcePull from OpenShift induced builds (vs. s2i command) ForcePull test context  ... (9m42.32619836s)
2) Skipping SSH core dump, which is not implemented for --- PASS: TestExtended (582.46s)

The failure stems from some hiccups with rsync in copying the logs files.

Are we still holding on merge based on @mfojtik 's request ?

File: pkg/build/builder/sti.go [hunk 7]
```diff
@@ -159,6 +158,16 @@ func (s *STIBuilder) Build() error {
 		ContextDir: s.build.Spec.Source.ContextDir,
 	}
 
+	if s.build.Spec.Strategy.SourceStrategy.ForcePull {
+		glog.V(4).Infof("With force pull true, setting policies to %s", s2iapi.PullAlways)
+		config.PreviousImagePullPolicy = s2iapi.PullAlways
+		config.BuilderPullPolicy = s2iapi.PullAlways
+	} else {
+		glog.V(4).Infof("With force pull false, setting policies to %s", s2iapi.PullAlways)
```
Comment (bparees) [inline pkg/build/builder/sti.go]: PullIfNotPresent

Comment (gabemontero) [inline pkg/build/builder/sti.go]: ahh ... ok, just pushed

**Comment by bparees** [comment]
@mfojtik if you have comments we'll handle them in a follow up.
[merge]

**Comment by gabemontero** [comment]
hmmm - same scenario ... tests passed, rsync of logs failed ... also, the test infrastructure is trying to call forcepull.sh (and it appears as if the extended tests are still running with a forcepull focus ... perhaps the prior forcepull only run has left a bit set somewhere).  The message after only running the forcepull test:

 `Running TEST_REPORT_DIR=/tmp/openshift-extended-tests/junit/extended test/extended/forcepull.sh...`

forcepull.sh used to exist, but was deleted as part of one of the extended test refactors; i would have to dig some more to see if the missing forcepull.sh is related to the rsync failures, and if it makes sense to add a forcepull.sh back in.

I'm thinking of trying a full extended test run manually first (perhaps unset the run only forcepull bit if my theory is correct).

@bparees thoughts when you get the chance

**Comment by bparees** [comment]
@gabemontero it's trying to call forcepull.sh because of your comment above where you named it as the extended test to run. the bot aggregrates all the test tags.  delete that comment and it'll stop trying to call forcepull.sh.

**Comment by gabemontero** [comment]
ok ... so my theory was correct - didn't think about the delete comment though - thx

**Comment by gabemontero** [comment]
[test]

**Comment by bparees** [comment]
also the rsync of the logs didn't fail in general, it just didn't find certain directories because you didn't run those tests.  that isn't causing the build failure, the build failure is because forcepull.sh wasn't found, so that's a failure code.

So i think once you remove your extended:forcepull comment, and then add another test tag so it reruns, you'll get a clean run.

**Comment by bparees** [comment]
@gabemontero i'm not sure how test and testonlyextended interact since they are basically opposites.  you probably want:
|test||extended:core(forcepull)|  or you want |testonlyextended||extended:core(forcepull)|

(with brackets instead of pipes of course.  just trying to avoid messing up your PR)

**Comment by bparees** [comment]
or you can just watch it and see what happens and then we'll learn something.

**Comment by bparees** [comment]
[test]

**Pull request merged**